eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala

```diff
@@ -0,0 +1,172 @@
+ * This should never be used alone but can be xor-ed with the OS random number generator in case it completely breaks.
+ */
+private class WeakRandom() extends RandomGenerator {
+
+  private val stream = new ChaCha7539Engine()
+  private val seed = new Array[Byte](32)
+  private var lastByte: Byte = 0
+  private var opsSinceLastSample: Int = 0
+
+  private val memoryMXBean = ManagementFactory.getMemoryMXBean
+  private val runtimeMXBean = ManagementFactory.getRuntimeMXBean
+  private val threadMXBean = ManagementFactory.getThreadMXBean
+
+  // sample some initial entropy
+  sampleEntropy()
+
+  private def feedDigest(sha: SHA256Digest, i: Int): Unit = {
+    sha.update(i.toByte)
+    sha.update((i >> 8).toByte)
+    sha.update((i >> 16).toByte)
+    sha.update((i >> 24).toByte)
+  }
+
+  private def feedDigest(sha: SHA256Digest, l: Long): Unit = {
+    sha.update(l.toByte)
+    sha.update((l >> 8).toByte)
+    sha.update((l >> 16).toByte)
+    sha.update((l >> 24).toByte)
+    sha.update((l >> 32).toByte)
+    sha.update((l >> 40).toByte)
+  }
+
+  /** The entropy pool is regularly enriched with newly sampled entropy. */
+  private def sampleEntropy(): Unit = {
+    opsSinceLastSample = 0
+
+    val sha = new SHA256Digest()
+    sha.update(seed, 0, 32)
+    feedDigest(sha, System.currentTimeMillis())
+    feedDigest(sha, System.identityHashCode(new Array[Int](1)))
+    feedDigest(sha, memoryMXBean.getHeapMemoryUsage.getUsed)
+    feedDigest(sha, memoryMXBean.getNonHeapMemoryUsage.getUsed)
+    feedDigest(sha, runtimeMXBean.getPid)
+    feedDigest(sha, runtimeMXBean.getUptime)
+    feedDigest(sha, threadMXBean.getCurrentThreadCpuTime)
+    feedDigest(sha, threadMXBean.getCurrentThreadUserTime)
+    feedDigest(sha, threadMXBean.getPeakThreadCount)
+
+    sha.doFinal(seed, 0)
+    // NB: init internally resets the engine, no need to reset it explicitly ourselves.
+    stream.init(true, new ParametersWithIV(new KeyParameter(seed), new Array[Byte](12)))
+  }
+
+  /** We sample new entropy approximately every 32 operations and at most every 64 operations. */
+  private def shouldSample(): Boolean = {
+    opsSinceLastSample += 1
+    val condition1 = -4 <= lastByte && lastByte <= 4
+    val condition2 = opsSinceLastSample >= 64
+    condition1 || condition2
+  }
+
+  def addEntropy(entropy: Array[Byte]): Unit = synchronized {
+    if (entropy.nonEmpty) {
+      val sha = new SHA256Digest()
+      sha.update(seed, 0, 32)
+      sha.update(entropy, 0, entropy.length)
+      sha.doFinal(seed, 0)
+      // NB: init internally resets the engine, no need to reset it explicitly ourselves.
+      stream.init(true, new ParametersWithIV(new KeyParameter(seed), new Array[Byte](12)))
+    }
+  }
+
+  def nextBytes(bytes: Array[Byte]): Unit = synchronized {
+    if (shouldSample()) {
+      sampleEntropy()
+    }
+    stream.processBytes(bytes, 0, bytes.length, bytes, 0)
```

Comment:
Is it safe to use `bytes` here as both input and output to  `processBytes()` ?

Reply:
It should be how we do in-place stream encryption (without needing to allocate a new array for no reason).
But this is definitely an area where you should double-check that I didn't miss something!

Reply:
I can add a test vector to verify that this does what we expect.

Reply:
I added a test case in https://github.com/ACINQ/eclair/pull/1774/commits/9fb8417d1b8b85a9c63904768f599a5ad82fb5e2 to prove this works as expected.